src/Analyzer/Resolve/IdentifierResolver.cpp

```diff
@@ -407,20 +407,39 @@ QueryTreeNodePtr IdentifierResolver::tryResolveTableIdentifierFromDatabaseCatalo
     }
 
     StorageID storage_id(database_name, table_name);
-    storage_id = context->resolveStorageID(storage_id);
+    storage_id = context->resolveStorageID(storage_id, Context::ResolveAll, /*get_uuid*/ false);
     bool is_temporary_table = storage_id.getDatabaseName() == DatabaseCatalog::TEMPORARY_DATABASE;
 
     StoragePtr storage;
+    TableLockHolder storage_lock;
 
-    if (is_temporary_table)
-        storage = DatabaseCatalog::instance().getTable(storage_id, context);
-    else
-        storage = DatabaseCatalog::instance().tryGetTable(storage_id, context);
+    // Find the table and lock it. Retry in case the table is dropped in the brief time before we lock it.
+    // In particular, this makes sure `SELECT FROM a` doesn't spuriously fail while another thread
+    // does `EXCHANGE TABLES a AND temp; DROP TABLE temp;`. Important for refreshable materialized views.
```

Comment:
Refreshable MVs should rely on UUIDs to avoid races with EXCHANGE TABLES and DROP/CREATE